src/gpio.cpp

```diff
@@ -0,0 +1,31 @@
+/**
+ * @file gpio.cpp
+ *
+ * @brief HAL GPIO class source
+ *
+ * @date 09/2024
+ */
+
+#include "gpio.hpp"
+
+namespace hal {
+Gpio::Gpio(const Config& config) : port{config.port}, pin{config.pin}, active_state{config.active_state} { }
+
+bool Gpio::read() const {
+    return HAL_GPIO_ReadPin(this->port, this->pin) == this->active_state;
+}
+
+void Gpio::write(bool state) {
+    GPIO_PinState pin_state = this->active_state;
+
+    if (!state) {
+        pin_state = (this->active_state == GPIO_PIN_SET) ? GPIO_PIN_RESET : GPIO_PIN_SET;
+    }
+
+    HAL_GPIO_WritePin(this->port, this->pin, pin_state);
+}
```

Comment:
```suggestion
void Gpio::write(bool state) {
    GPIO_PinState pin_state = state ? this->active_state : static_cast<GPIO_PinState>(!this->active_state);
    HAL_GPIO_WritePin(this->port, this->pin, pin_state);
}
```

Assim nao fica mais bonitinho? kkkk

Reply:
assim da `Implicit conversion 'GPIO_PinState' -> 'bool'`, pq o `this->active_state` eh um `GPIO_PinState` asuihfaiufhas

eu nao gosto do jeito que ta tambem, mas passei um tempinho brigando com o linter

talvez uma abordagem boa seja apostar na legibilidade e meter um nolint mesmo

Reply:
da pra usar umas funções privadas pra deixar mais legivel. tipo algo assim:

```cpp
void Gpio::write(bool state) {
    GPIO_PinState pin_state = get_pin_state_from_bool(state);
    HAL_GPIO_WritePin(this->port, this->pin, pin_state);
}

GPIO_PinState Gpio::get_pin_state_from_bool(bool state) const {
    return (state == is_active_high()) ? GPIO_PIN_SET : GPIO_PIN_RESET;
}

bool Gpio::is_active_high() const {
    return this->active_state == GPIO_PIN_SET;
}
```

sei la kkkkk

Reply:
Pq q só n faz uma conversão explícita com static_cast?

Reply:
tbm da

Reply:
`static_cast<GPIO_PinState>(!static_cast<bool>(this->active_state))`

nao é acoisa mais linda do mundo kkkk

Reply:
Uai, se ele tá reclamando de implicit conversioin aí é só fazer uma explicit conversion asuhasuhasuhsa

Reply:
> Uai, se ele tá reclamando de implicit conversioin aí é só fazer uma explicit conversion asuhasuhasuhsa

sim po, mas fica feio aaoskfpasfop

Reply:
static_cast é pros feios, os bonitos fazem `bool(this->active_state)` 😘